### PR TITLE
Access traits for predicates

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

**Comment by dalg24** [comment]
This is ready for review.

@aprokop I compared on-node performance with 5589972 and there was no noticeable effect.
I ran 1M values and 100k queries.

File: src/ArborX_LinearBVH.hpp [hunk 1]
```diff
@@ -56,9 +56,13 @@ class BoundingVolumeHierarchy
   template <typename Predicates, typename... Args>
   void query(Predicates const &predicates, Args &&... args) const
   {
-    // FIXME lame placeholder for concept check
-    static_assert(Kokkos::is_view<Predicates>::value, "must pass a view");
-    using Tag = typename Predicates::value_type::Tag;
+    // FIXME placeholder for concept check
```
Comment (sslattery) [inline src/ArborX_LinearBVH.hpp]: Are you going to keep this comment?

File: src/details/ArborX_DetailsBatchedQueries.hpp [hunk 2]
```diff
@@ -82,18 +84,30 @@ struct BatchedQueries
     return sortObjects(morton_codes);
   }
 
-  template <typename T>
-  static Kokkos::View<T *, DeviceType>
-  applyPermutation(Kokkos::View<size_t const *, DeviceType> permute,
-                   Kokkos::View<T *, DeviceType> v)
+  // NOTE  trailing return type seems required :(
+  // error: The enclosing parent function ("applyPermutation") for an extended
```
Comment (sslattery) [inline src/details/ArborX_DetailsBatchedQueries.hpp]: Is this an NVCC limitation? I would assume so.

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp [hunk 3]
```diff
@@ -95,16 +93,18 @@ void BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
...
   Kokkos::Profiling::pushRegion("ArborX:BVH:sort_queries");
 
   auto const permute =
       Details::BatchedQueries<DeviceType>::sortQueriesAlongZOrderCurve(
-          bvh.bounds(), queries);
+          bvh.bounds(), predicates);
 
-  queries =
-      Details::BatchedQueries<DeviceType>::applyPermutation(permute, queries);
+  // FIXME  readability!  queries is a sorted copy of the predicates
```
Comment (sslattery) [inline src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: What is the readability issue here?

File: src/details/ArborX_DetailsTags.hpp [hunk 4]
```diff
@@ -45,6 +45,12 @@ struct Tag<Box>
   using type = BoxTag;
 };
 
+template <typename T>
```
Comment (sslattery) [inline src/details/ArborX_DetailsTags.hpp]: So does this let you just add a `Tag` typedef to any object and then you can extract the tag without having to write a specialization for that object?

File: src/details/ArborX_DetailsTags.hpp [hunk 5]
```diff
@@ -45,6 +45,12 @@ struct Tag<Box>
...
+struct Tag<T>
```
Comment (sslattery) [inline src/details/ArborX_DetailsTags.hpp]: Also are their other types of tags you would expect? The name seems pretty generic. This is OK unless it would be possible for a type to have more than one `Tag` depending on how it is used.

File: src/details/ArborX_DetailsTreeConstruction.hpp [hunk 6]
```diff
@@ -36,6 +36,19 @@ namespace ArborX
 {
 namespace Details
 {
+
+template <typename T, typename TTag>
+struct TagHelper
+{
+private:
+  using accessor_return_type =
```
Comment (sslattery) [inline src/details/ArborX_DetailsTreeConstruction.hpp]: man I love meta programming

File: src/ArborX_LinearBVH.hpp (hunk 1 above)
Comment (dalg24) [reply src/ArborX_LinearBVH.hpp]: Yes concept check coming up next.

File: src/details/ArborX_DetailsBatchedQueries.hpp (hunk 2 above)
Comment (dalg24) [reply src/details/ArborX_DetailsBatchedQueries.hpp]: Yep.

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp (hunk 3 above)
Comment (dalg24) [reply src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: Function argument is `predicates` and it is some unspecified type that is only required to have a valid matching specialization of the access traits and bam next line I create `queries` which is a Kokkos view and keep going as before.

File: src/details/ArborX_DetailsTags.hpp (hunk 4 above)
Comment (dalg24) [reply src/details/ArborX_DetailsTags.hpp]: Correct.  I plan on changing the case `Tag` -> `tag` sometime soon for consistency.

File: src/details/ArborX_DetailsTreeConstruction.hpp [hunk 7]
```diff
@@ -36,6 +36,19 @@ namespace ArborX
...
+      std::decay_t<decltype(Traits::Access<T, TTag>::get(
+          std::declval<T const &>(), std::declval<int>()))>;
```
Comment (aprokop) [inline src/details/ArborX_DetailsTreeConstruction.hpp]: Why do you need the second `declval`?

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp [hunk 8]
```diff
@@ -95,16 +93,18 @@ void BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
...
+  auto queries = Details::BatchedQueries<DeviceType>::applyPermutation(
+      permute, predicates);
```
Comment (masterleinad) [inline src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: So this doesn't impact performance?

File: src/details/ArborX_DetailsTags.hpp [hunk 9]
```diff
@@ -28,7 +28,7 @@ struct BoxTag
 {
 };
 
-template <typename Geometry>
+template <typename T, typename Enable = void>
```
Comment (masterleinad) [inline src/details/ArborX_DetailsTags.hpp]: Wouldn't it make sense to separate geometry tags and predicate tags? Is there any function that is supposed to handle both?

File: src/details/ArborX_DetailsTags.hpp (hunk 5 above)
Comment (masterleinad) [inline src/details/ArborX_DetailsTags.hpp]: Hmm... It looks a little bit weird to treat `Point` and `Box` separately here. I would be happier if `Point` and `Box` would also have a `Tag` alias and we would not have to specialize here.

File: src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp (hunk 8 above)
Comment (dalg24) [reply src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp]: This does not change the algorithm and the choices that were previously done, which was to make a sorted copy of the queries/predicates instead of accessing with indirection.  I had also explored populating the results in permuted order and reverse the permutation at the end but that was slower.
We can try to not make a copy (i.e. drop `queries`) and do `Access::get(predicates, permute(i))` but I don't want to explore this in this PR.  Also I am pretty sure I had explored that option back then.

File: src/details/ArborX_DetailsTags.hpp (hunk 9 above)
Comment (dalg24) [reply src/details/ArborX_DetailsTags.hpp]: And put them what header instead?  This is good enough.

File: src/details/ArborX_DetailsTags.hpp (hunk 5 above)
Comment (dalg24) [reply src/details/ArborX_DetailsTags.hpp]: Not going to happen here.

File: src/ArborX_LinearBVH.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply src/ArborX_LinearBVH.hpp]: See #113 
Would check that matching specialization exist, that it has the proper `size()` and `get()` static member functions, the `memory_space` member type, that the tag can be deduced for the dispatch, etc.
Without it the user gets error messages that are very hard to parse.

File: src/ArborX_LinearBVH.hpp [hunk 10]
```diff
@@ -56,9 +56,13 @@ class BoundingVolumeHierarchy
...
+
+    using Tag =
+        typename Details::TagHelper<Predicates, Traits::PredicatesTag>::type;
+    static_assert(std::is_same<Tag, Details::NearestPredicateTag>::value ||
+                      std::is_same<Tag, Details::SpatialPredicateTag>::value,
+                  "Invalid tag for the predicates");
```
[the remark this answers is not in the document]
Comment (dalg24) [reply src/ArborX_LinearBVH.hpp]: Won't fix.  Will be superseded by #113

File: src/details/ArborX_DetailsTreeConstruction.hpp [hunk 11]
```diff
@@ -36,6 +36,19 @@ namespace ArborX
 {
 namespace Details
 {
+
+template <typename T, typename TTag>
```
[the remark this answers is not in the document]
Comment (dalg24) [reply src/details/ArborX_DetailsTreeConstruction.hpp]: No that's not what this is for.  It is there to deduce the tag from the return type of the `get()` access traits.

File: src/details/ArborX_DetailsTreeConstruction.hpp (hunk 7 above)
Comment (dalg24) [reply src/details/ArborX_DetailsTreeConstruction.hpp]: Could probably replace with an integer but the second argument is needed because `get` takes two params.

File: src/details/ArborX_DetailsTreeConstruction.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply src/details/ArborX_DetailsTreeConstruction.hpp]: ... which is what I said in the last sentence.

**Comment by dalg24** [comment]
> Overall, it looks really good. There are two points I want to bring up:
> 
> 1. The words predicates and primitives are really hard for me. It always takes me several seconds to remember what they are, so personally I would prefer something simpler. Am I the only one with this issue?
> 2. I'm not sure how `Primitives` `DataAccess` interacts with `BoxTag` and `PointTag`. Am I correct in understanding that the current approach does not allow for the fine search to be implemented at the same time?

1.  What other terminology would you like to suggest?
2. Not necessarily.  Storing user geometries instead of boxes in the leaf nodes would require non trivial changes so the geometry tag needed at construction is not a big concern at the moment.  I would approach the fine search problem by "piggy packing" some work when collision are found.  My current thinking is I will solve it via some new access traits for the results, with a specialization for a new second template argument `ResultsTag`.
